kbfsedits/tlf_history.go

```diff
@@ -0,0 +1,354 @@
+		// Disregard any file that's already been deleted.
+		if hasEvent && event.delete {
+			return false
+		}
+
+		// We only care about files, so skip dir and sym creates.
+		if notification.FileType != EntryTypeFile {
+			return false
+		}
+
+		// We only need one modify message per writer per file.
+		if r.modifiedFiles[writer][filename] {
+			return false
+		}
+
+		wn, ok := r.byWriter[writer]
+		if !ok {
+			wn = &writerNotifications{writer, nil}
+			r.byWriter[writer] = wn
+		}
+		wn.writerName = writer
+		wn.notifications = append(wn.notifications, notification)
+
+		modified, ok := r.modifiedFiles[writer]
+		if !ok {
+			modified = make(map[string]bool)
+			r.modifiedFiles[writer] = modified
+		}
+		modified[filename] = true
+
+		if len(wn.notifications) == maxEditsPerWriter {
+			// We have enough edits for this user.
+			return true
+		}
+	case NotificationRename:
+		// If the file already has a final event, move that to the old
+		// filename.  Otherwise, this is the final event.
+		if hasEvent {
+			r.fileEvents[notification.Params.OldFilename] = event
+			delete(r.fileEvents, eventFilename)
+		} else {
+			r.fileEvents[notification.Params.OldFilename] =
+				fileEvent{newName: eventFilename}
+		}
+
+		// The renamed file overwrote any existing file with the new
+		// name.
+		r.fileEvents[eventFilename] = fileEvent{delete: true}
+	case NotificationDelete:
+		r.fileEvents[eventFilename] = fileEvent{delete: true}
+	}
+	return false
+}
+
+func (th *TlfHistory) recomputeLocked() (
+	history writersByRevision, writersWhoNeedMore map[string]bool) {
+	writersWhoNeedMore = make(map[string]bool)
+
+	// Copy the writer notifications into a heap.
+	var writersHeap writersByRevision
+	for _, wn := range th.byWriter {
+		wnCopy := writerNotifications{
+			writerName:    wn.writerName,
+			notifications: make(notificationsByRevision, len(wn.notifications)),
+		}
+		copy(wnCopy.notifications, wn.notifications)
+		writersHeap = append(writersHeap, &wnCopy)
+	}
+	heap.Init(&writersHeap)
+
+	r := newRecomputer()
+
+	// Iterate through the heap.  The writer with the next highest
+	// revision will always be at index 0.  Process that writer's
+	// first notification, then remove it and fix the heap so that the
+	// next highest revision is at index 0.  That way events that
+	// happen more recently (like deletes and renames) can be taken
+	// into account when looking at older events.
+	for writersHeap.Len() > 0 {
+		nextWriter := writersHeap[0].writerName
+		nextNotification := writersHeap[0].notifications[0]
+		doTrim := r.processNotification(nextWriter, nextNotification)
+
+		// Remove that notification, and fix the heap because this
+		// writer has a different newest revision.
+		if doTrim {
+			// Trim all earlier revisions because they won't be needed
+			// for the cached history.
+			numProcessed := r.numProcessed[nextWriter]
+			th.byWriter[nextWriter].notifications =
+				th.byWriter[nextWriter].notifications[numProcessed:]
```

Comment:
Is this essentially to make `th.byWriter[nextWriter].notifications` an empty slice? If so, why not just make it `nil`?

Reply:
No, `numProcessed` could be more than 0 (and will be, in the common case).  So this is removing items from the end of the slice that were not used in the making of the history, not necessarily making it `nil`.

Reply:
Actually this might be backwards, this seems to be removing items from the beginning of the slice, the ones that were actually processed.  I'm going to try to fix that and add a test.

Reply:
> No, numProcessed could be more than 0 (and will be, in the common case).

By reading the code, I was under the impression that `numProcessed` will always be equal to `len(th.byWriter[nextWriter].notifications)`, making this an empty slice. (Since the `numProcessed` marks the start of the slicing).

Reply:
It's not, because of this: https://github.com/keybase/kbfs/blob/15309b94d93e4d6d3b900e0b0d5374ca53ef7749/kbfsedits/tlf_history.go#L232-L234

When we get enough notifications for the user, we return `doTrim=true`, which causes the caller to truncate the rest of the notifications away from the slice.  Or that's what it will do, with my new commit.